File: app/src/main/java/com/forkmang/adapter/Walkin_listing_Adapter.java
package com.forkmang.adapter;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.forkmang.R;
import com.forkmang.data.RestoData;
import com.forkmang.fragment.Walkin_detail_Fragment;

import java.util.ArrayList;


public class Walkin_listing_Adapter extends RecyclerView.Adapter<Walkin_listing_Adapter.RestoListItemHolder> {
    Activity activity;
    ArrayList<RestoData> resto_dataArrayList;
    String coming_from;
    Context ctx;

    public Walkin_listing_Adapter(Activity activity, ArrayList<RestoData> bookTable_dataArrayList,String coming_from, Context ctx) {
        this.activity = activity;
        this.resto_dataArrayList = bookTable_dataArrayList;
        this.coming_from = coming_from;
        this.ctx=ctx;
    }

    public Walkin_listing_Adapter(Activity activity) {
        this.activity = activity;
    }

    /*book_table_cell*/
    @NonNull
    @Override
    public RestoListItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.book_table_cell, null);
        RestoListItemHolder bookTableItemHolder = new RestoListItemHolder(v);
        return bookTableItemHolder;
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull RestoListItemHolder holder, int position) {
        RestoData bookTable = resto_dataArrayList.get(position);
        holder.txtrestroname.setText(bookTable.getRest_name());
        holder.txt_endtime.setText(bookTable.getEndtime());
        holder.txttotalkm.setText(bookTable.getDistance()+" Km");
    }

    @Override
    public int getItemCount(){
       return resto_dataArrayList.size();
    }

    public class RestoListItemHolder extends RecyclerView.ViewHolder {
        ImageView imgproduct;
        TextView txtrestroname, txt_endtime, txt_distance,txttotalkm,txt_ratingno;
        RatingBar rating_bar;
        RelativeLayout relative_view;
        public RestoListItemHolder(@NonNull View itemView) {
            super(itemView);
            relative_view = itemView.findViewById(R.id.relative_view);
            txtrestroname=itemView.findViewById(R.id.txtrestroname);
            txt_endtime=itemView.findViewById(R.id.txt_endtime);
            txt_distance=itemView.findViewById(R.id.txt_distance);
            txttotalkm=itemView.findViewById(R.id.txt_totalkm);
            txt_ratingno=itemView.findViewById(R.id.txt_ratingno);
            imgproduct=itemView.findViewById(R.id.imgrestro);
            rating_bar=itemView.findViewById(R.id.rating_bar);

            relative_view.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                RestoData restoData = resto_dataArrayList.get(position);
                String resturant_id = restoData.getId();

                if(coming_from.equalsIgnoreCase("listing"))
                {
                    Intent intent = new Intent(activity, Walkin_detail_Fragment.class);
                    intent.putExtra("resturant_id", resturant_id);
                    intent.putExtra("restromodel", restoData);
                    activity.startActivity(intent);
                }
                else{
                      ((Walkin_detail_Fragment) ctx).callapi_getquess(resturant_id);
                }
                //activity.finish();
            });

        }
    }



}

File: app/src/main/java/com/forkmang/adapter/WalkinListingAdapter.kt
package com.forkmang.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.forkmang.R
import com.forkmang.data.RestoData

class WalkinListingAdapter(
    var coming_from: String?,
    private val ItemOnClick: ((restId: String, restroData: RestoData?) -> Unit)
) : ListAdapter<RestoData, WalkinListingAdapter.RestoListItemHolder>(ItemCallback) {

    object ItemCallback : DiffUtil.ItemCallback<RestoData>() {
        override fun areItemsTheSame(oldItem: RestoData, newItem: RestoData): Boolean =
            oldItem.id == newItem.id

        @SuppressLint("DiffUtilEquals")
        override fun areContentsTheSame(oldItem: RestoData, newItem: RestoData): Boolean =
            oldItem == newItem
    }

    var resto_dataArrayList: List<RestoData> = emptyList()
        set(value) {
            field = value
            submitList(field)
        }

    /*book_table_cell*/
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RestoListItemHolder {
        val v =
            LayoutInflater.from(parent.context).inflate(R.layout.book_table_cell, null)
        return RestoListItemHolder(v)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: RestoListItemHolder, position: Int) {
        val bookTable = resto_dataArrayList[position]
        holder.txtrestroname.text = bookTable.rest_name
        holder.txtEndTime.text = bookTable.endtime
        holder.txttotalkm.text = bookTable.distance + " Km"
    }


    inner class RestoListItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var imgproduct: ImageView
        var txtrestroname: TextView
        var txtEndTime: TextView
        var txtDistance: TextView
        var txttotalkm: TextView
        var txtRatingNo: TextView
        var ratingBar: RatingBar
        var relativeView: RelativeLayout

        init {
            relativeView = itemView.findViewById(R.id.relative_view)
            txtrestroname = itemView.findViewById(R.id.txtrestroname)
            txtEndTime = itemView.findViewById(R.id.txt_endtime)
            txtDistance = itemView.findViewById(R.id.txt_distance)
            txttotalkm = itemView.findViewById(R.id.txt_totalkm)
            txtRatingNo = itemView.findViewById(R.id.txt_ratingno)
            imgproduct = itemView.findViewById(R.id.imgrestro)
            ratingBar = itemView.findViewById(R.id.rating_bar)
            relativeView.setOnClickListener {
                val position = bindingAdapterPosition
                val restoData = resto_dataArrayList[position]
                val resturantId = restoData.id
                if (coming_from.equals("listing", ignoreCase = true)) {
                    ItemOnClick(resturantId ?: "", restoData)
                } else {
                    ItemOnClick(resturantId ?: "", null)
                }
            }
        }
    }
}